### PR TITLE
Add Gridsampler contrib op

### PR DESCRIPTION
**Pull request by hwangdeyu**
**Description**: 
Implement contrib op Gridsampler 
Given an `input` and a flow-field `grid`, computes the `output` using `input` values and pixel locations from `grid`.

**Motivation and Context**
Provide reference to implement the kernal.
This is an important op which is used in vision related models.
It's often used in conjunction with affine_grid() to build Spatial Transformer Networks.

### CODE REVIEW

File: onnxruntime/test/python/contrib_ops/onnx_test_gridsampler.py [hunk 1]
```diff
@@ -0,0 +1,533 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (askhade) [inline onnxruntime/test/python/contrib_ops/onnx_test_gridsampler.py:1]: For ORT just the unit tests in gridsampler_test.cc should be enough... these node tests should be added to ONNX PR...

**Review by XiyinOSS — COMMENTED** [review]
Change floating literals from double to float
e.g. 0.0 ->0.f

File: onnxruntime/test/python/contrib_ops/onnx_test_gridsampler.py (hunk 1 above)
Comment (hwangdeyu) [reply onnxruntime/test/python/contrib_ops/onnx_test_gridsampler.py:1]: ok, I also see a few other ops added the onnx test, wiil remove it.

**Comment by hwangdeyu** [comment]
Cause I don't provider the kernal, the some of CI tests will be failed.
@askhade  Any more comments? Thanks.

File: onnxruntime/test/contrib_ops/gridsampler_test.cc [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  OpTester test("GridSampler", 1, kMSDomain);
+  test.AddInput<float>("X", {1, 1, 3, 2}, {0.0f, 1.0f, 2.0f, 3.0f, 4.0f, 5.0f});
+  test.AddInput<float>("Grid", {1, 2, 4, 2},
+    {-10.0000f, -10.0000f, -5.0000f, -5.0000f,
+     -0.2000f, -0.2000f, 10.0000f, 10.0000f,
+     10.0000f, 10.0000f, -0.2000f, -0.2000f,
+     5.0000f, 5.0000f, 10.0000f, 10.0000f
+    });
+  test.AddAttribute("padding_mode", "border");
+  test.AddOutput<float>("Y", {1, 1, 2, 4}, {5.0000f, 0.0000f, 1.7000f, 5.0000f, 5.0000f, 1.7000f, 5.0000f, 5.0000f});
```
[the remark this answers is not in the document]
Comment (hwangdeyu) [reply onnxruntime/test/contrib_ops/gridsampler_test.cc:64]: The reproduction of this example output as follow, seems 5.f is the result?
```
import torch
X = torch.arange(2 * 3).view(1, 1, 3, 2).float()
grid = torch.tensor([[10,10], [-5, -5], [-0.2,-0.2], [10, 10],[10, 10], [-0.2, -0.2], [5,5], [10, 10]]).view(1, 2, 4, 2).float()
torch.nn.functional.grid_sample(input, grid,padding_mode="border")

>>tensor([[[[5.0000, 0.0000, 1.7000, 5.0000],
          [5.0000, 1.7000, 5.0000, 5.0000]]]])
```

[the remark this answers is not in the document]
Comment (XiyinOSS) [reply onnxruntime/test/contrib_ops/gridsampler_test.cc:64]: The first grid point is different in your script than in the code
grid = torch.tensor([[**10,10**]
vs
test.AddInput<float>("Grid", {1, 2, 4, 2},
    **{-10.0000f, -10.0000f,**

[the remark this answers is not in the document]
Comment (hwangdeyu) [reply onnxruntime/test/contrib_ops/gridsampler_test.cc]: Thanks for the comment, fixed it and checked all the tests again.

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 3]
```diff
@@ -2964,7 +2964,96 @@ It's an extension of Gelu. It takes the sum of input A and bias input B as the i
           ctx.getOutputType(0)
               ->CopyFrom(input_type->optional_type().elem_type());
           });
-  
+
+  static const char* GridSampler_ver1_doc = R"DOC(
+      Given an `input` and a flow-field `grid`, computes the `output` using `input` values and pixel locations from `grid`.
+      Currently, only spatial (4-D) inputs are supported. For `input` with shap (N, C, H, W) and `grid` with shape (N, H_out, W_out, 2),
```
Comment (askhade) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2970]: typo: shap -> shape

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 4]
```diff
@@ -2964,7 +2964,96 @@ It's an extension of Gelu. It takes the sum of input A and bias input B as the i
...
+      the `output` will have shape (N, C, H_out, W_out).
+      For each output location `output[n, :, h, w]`, the size-2 vector `grid[n, h, w]` specifies `input` pixel locations `x` and `y`,
+      which are used to interpolate the output value `output[n, :, h, w]`.
+      The GridSampler operator is often used in conjunction with affine_grid doing grid generator
```
Comment (askhade) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2974]: there is no affine_grid operator in ONNX or ORT contrib ops... do we need to add this op too?

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 5]
```diff
@@ -2964,7 +2964,96 @@ It's an extension of Gelu. It takes the sum of input A and bias input B as the i
           ctx.getOutputType(0)
               ->CopyFrom(input_type->optional_type().elem_type());
           });
-  
+
+  static const char* GridSampler_ver1_doc = R"DOC(
```
Comment (askhade) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2968]: should we name this GridSample as opposed to GridSampler

**Comment by askhade** [comment]
@hwangdeyu : you need to disable the unit tests until a kernel is added for this OP ... this should fix the CIs

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 5 above)
Comment (hwangdeyu) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2968]: In PyTorch, the implementation name is GridSampler. Is better that we name GridSample instand of Gridsampler？ 
https://github.com/pytorch/pytorch/issues/41957
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/GridSampler.h

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 4 above)
Comment (hwangdeyu) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2974]: The affine_grid is used to create the `Grid` input of grid_sample.
The github community and fomm model only request for `grid_sampler`, so I think we don't need this op now.

**Review by askhade — APPROVED** [review]

File: docs/ContribOperators.md [hunk 6]
```diff
@@ -1185,6 +1186,58 @@ This version of the operator has been available since version 1 of the 'com.micr
 </dl>
 
 
+### <a name="com.microsoft.GridSample"></a><a name="com.microsoft.gridsample">**com.microsoft.GridSample**</a>
+
+  Given an `input` and a flow-field `grid`, computes the `output` using `input` values and pixel locations from `grid`.
+        Currently, only spatial (4-D) inputs are supported. For `input` with shape (N, C, H, W) and `grid` with shape (N, H_out, W_out, 2),
+        the `output` will have shape (N, C, H_out, W_out).
+        For each output location `output[n, :, h, w]`, the size-2 vector `grid[n, h, w]` specifies `input` pixel locations `x` and `y`,
```
Comment (BowenBao) [inline docs/ContribOperators.md:1194]: in PyTorch grid_sample, the input/output shape could be different: `input[n, c, h_in, w_in]`, `output[n, c, h_out, w_out]`. 

https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html#torch.nn.functional.grid_sample

Is it intended to design contrib GridSample having same shape for input and output?

Comment (hwangdeyu) [reply docs/ContribOperators.md:1194]: yes, it's the same with `output` and `grid`. The description may be not very clear. I will amend this description.

**Comment by timiil** [comment]
hi, any update for this ? we are very need this op ...

```
System.InvalidOperationException: Error initializing model :Microsoft.ML.OnnxRuntime.OnnxRuntimeException: [ErrorCode:Fail] Load model from D:\code\ml\samples\csharp\getting-started\DeepLearning_ObjectDetection_Onnx\ObjectDetectionConsoleApp\bin\Debug\net5\../../../assets\Model\mask_rcnn_r50_caffe_fpn_mstrain-poly_1x_slock1_cpu.onnx failed:Fatal error: grid_sampler is not a registered function/op
   at Microsoft.ML.OnnxRuntime.NativeApiStatus.VerifySuccess(IntPtr nativeStatus)
   at Microsoft.ML.OnnxRuntime.InferenceSession.Init(String modelPath, SessionOptions options, PrePackedWeightsContainer prepackedWeightsContainer)
   at Microsoft.ML.OnnxRuntime.InferenceSession..ctor(String modelPath)
   at Microsoft.ML.Transforms.Onnx.OnnxModel..ctor(String modelFile, Nullable`1 gpuDeviceId, Boolean fallbackToCpu, Boolean ownModelFile, IDictionary`2 shapeDictionary, Int32 recursionLimit)

```

**Comment by hwangdeyu** [comment]
> hi, any update for this ? we are very need this op ...
> 
> ```
> System.InvalidOperationException: Error initializing model :Microsoft.ML.OnnxRuntime.OnnxRuntimeException: [ErrorCode:Fail] Load model from D:\code\ml\samples\csharp\getting-started\DeepLearning_ObjectDetection_Onnx\ObjectDetectionConsoleApp\bin\Debug\net5\../../../assets\Model\mask_rcnn_r50_caffe_fpn_mstrain-poly_1x_slock1_cpu.onnx failed:Fatal error: grid_sampler is not a registered function/op
>    at Microsoft.ML.OnnxRuntime.NativeApiStatus.VerifySuccess(IntPtr nativeStatus)
>    at Microsoft.ML.OnnxRuntime.InferenceSession.Init(String modelPath, SessionOptions options, PrePackedWeightsContainer prepackedWeightsContainer)
>    at Microsoft.ML.OnnxRuntime.InferenceSession..ctor(String modelPath)
>    at Microsoft.ML.Transforms.Onnx.OnnxModel..ctor(String modelFile, Nullable`1 gpuDeviceId, Boolean fallbackToCpu, Boolean ownModelFile, IDictionary`2 shapeDictionary, Int32 recursionLimit)
> ```

Thanks for attention.
Yes, This PR is only to defind the `gridsample` op. If you want to use it as an contrib op please refer to the kernal inplemenation PR https://github.com/microsoft/onnxruntime/pull/8551.
By the way, the op name is `gridsample` instand of `grid_sampler`...

**Comment by timiil** [comment]
> 
> 
> > hi, any update for this ? we are very need this op ...
> > ```
> > System.InvalidOperationException: Error initializing model :Microsoft.ML.OnnxRuntime.OnnxRuntimeException: [ErrorCode:Fail] Load model from D:\code\ml\samples\csharp\getting-started\DeepLearning_ObjectDetection_Onnx\ObjectDetectionConsoleApp\bin\Debug\net5\../../../assets\Model\mask_rcnn_r50_caffe_fpn_mstrain-poly_1x_slock1_cpu.onnx failed:Fatal error: grid_sampler is not a registered function/op
> >    at Microsoft.ML.OnnxRuntime.NativeApiStatus.VerifySuccess(IntPtr nativeStatus)
> >    at Microsoft.ML.OnnxRuntime.InferenceSession.Init(String modelPath, SessionOptions options, PrePackedWeightsContainer prepackedWeightsContainer)
> >    at Microsoft.ML.OnnxRuntime.InferenceSession..ctor(String modelPath)
> >    at Microsoft.ML.Transforms.Onnx.OnnxModel..ctor(String modelFile, Nullable`1 gpuDeviceId, Boolean fallbackToCpu, Boolean ownModelFile, IDictionary`2 shapeDictionary, Int32 recursionLimit)
> > ```
> 
> Thanks for attention.
> Yes, This PR is only to defind the `gridsample` op. If you want to use it as an contrib op please refer to the kernal inplemenation PR #8551.
> By the way, the op name is `gridsample` instand of `grid_sampler`...

thanks for reply,   the name of 'grid_sampler' op is coming from the OpenMMLab -> mask-rcnn -> pytorch , the OpenMMLab tool convert the model into ONNX. Just FYI ， thanks you again
